### PR TITLE
Make tool, fully CLI and fully configurable by ENV

### PR DESCRIPTION
**Pull request by Moejoe90**
Refactored Command.js code

### CODE REVIEW

File: src/command.js [hunk 1]
```diff
@@ -2,74 +2,115 @@ import program from 'commander';
...
-export default (filePath, lastBlockNumber) => {
-  const watchConfig = YAML.load(filePath);
-  program
-    .version('0.1.0')
-    .option('-a, --addresses <n>', 'List of address', list)
-    .option('-f, --from [n]', 'From block', typeof watchConfig.from !== 'undefined' ? watchConfig.from : defaultBlockNumber)
-    .option('-t, --to [n]', 'To block', typeof watchConfig.to !== 'undefined' ? watchConfig.to : defaultBlockNumber)
-    .option('-q, --quick [n]', 'Quick Mode', typeof watchConfig.quick !== 'undefined' ? watchConfig.quick : false)
-    .option('-s, --save-state [n]', 'Save state', typeof watchConfig.saveState !== 'undefined' ? watchConfig.saveState : null);
+export class Commands {
```
Comment (MostafaBalata) [inline src/command.js]: Why did you change it to class?

File: src/command.js [hunk 2]
```diff
@@ -2,74 +2,115 @@ import program from 'commander';
...
-  const saveStatePath = program.saveState;
-  let lastBlockNumberFilePath = null;
+  /**
+   * Handle input value priorities where ENV variable is higher in priority then config file
+   * then default value
+   * @param {string} envName
+   * @return value
+   */
+  static handelInputValues(envName, fileInst, defaultValue) {
+    return getEnv(envName) ? getEnv(envName) : fileInst || defaultValue;
```
Comment (MostafaBalata) [inline src/command.js]: as far as I understand, I see you changed the priorities to be:
1- .env file
2- .watch
3- default

The questions is why should I put the address in .env? why there's possibility to put the watchConfig.node in watch file?

I think this will conduct to miss understand between .env and .watch.

File: .env.example [hunk 3]
```diff
@@ -1,4 +1,10 @@
...
-RPC_URL = http://localhost:8545
-OUTPUT_TYPE = "graylog"
-ACCESS_TOKEN = ""
\ No newline at end of file
+LOG_LEVEL= 'debug'
+SOCKET_IO_PORT = 3030
+RPC_URL = "https://nd.neufund.org:2096"
+OUTPUT_TYPE = "terminal"
+ACCESS_TOKEN = ""
+SAVE_STATE = "true"
```
Comment (MostafaBalata) [inline .env.example]: I still don't understand the meaning of string including "true" is the double quotes are mistake, or it mean something?

second thing why did you move it into the .env file?

File: src/command.js [hunk 4]
```diff
@@ -2,74 +2,115 @@ import program from 'commander';
...
+    const watchConfig = watchPath ? YAML.load(watchPath) : YAML.load(watchingConfigPath);
+    program
+      .version('0.1.0')
+      .option('-a, --addresses [n]', 'List of address', list, Commands.handelInputValues('ADDRESSES', watchConfig.addresses, ''))
+      .option('-f, --from [n]', 'From block', Commands.handelInputValues('FROM_BLOCK', watchConfig.from, 0))
+      .option('-t, --to [n]', 'To block', Commands.handelInputValues('TO_BLOCK', watchConfig.to, defaultBlockNumber))
+      .option('-q, --quick [n]', 'Quick Mode', Commands.handelInputValues('QUICK_MODE', watchConfig.quick, true))
+      .option('-s, --save-state [n]', 'Save state', Commands.handelInputValues('SAVE_STATE', watchConfig.saveState, null))
+      .option('-n, --node-url [n]', 'Node address', Commands.handelInputValues('RPC_URL', watchConfig.node, 'http://localhost:8545'))
+      .option('-l, --log-level [n]', 'Log level', Commands.handelInputValues('LOG_LEVEL', watchConfig.log_level, 'info'))
```
Comment (MostafaBalata) [inline src/command.js]: `'http://localhost:8545'` shouldn't be hard coded here, it should be config variable.

File: src/command.js [hunk 5]
```diff
@@ -2,74 +2,115 @@ import program from 'commander';
...
+  constructor(watchPath, lastBlockNumber) {
+    const watchConfig = watchPath ? YAML.load(watchPath) : YAML.load(watchingConfigPath);
+    program
+      .version('0.1.0')
+      .option('-a, --addresses [n]', 'List of address', list, Commands.handelInputValues('ADDRESSES', watchConfig.addresses, ''))
+      .option('-f, --from [n]', 'From block', Commands.handelInputValues('FROM_BLOCK', watchConfig.from, 0))
+      .option('-t, --to [n]', 'To block', Commands.handelInputValues('TO_BLOCK', watchConfig.to, defaultBlockNumber))
+      .option('-q, --quick [n]', 'Quick Mode', Commands.handelInputValues('QUICK_MODE', watchConfig.quick, true))
+      .option('-s, --save-state [n]', 'Save state', Commands.handelInputValues('SAVE_STATE', watchConfig.saveState, null))
+      .option('-n, --node-url [n]', 'Node address', Commands.handelInputValues('RPC_URL', watchConfig.node, 'http://localhost:8545'))
```
Comment (MostafaBalata) [inline src/command.js]: 'http://localhost:8545' shouldn't be hard coded here, it should be config variable

File: src/command.js [hunk 6]
```diff
@@ -2,74 +2,115 @@ import program from 'commander';
...
+      .option('-l, --log-level [n]', 'Log level', Commands.handelInputValues('LOG_LEVEL', watchConfig.log_level, 'info'))
+      .option('-p, --socket-port [n]', 'IO socket port number', Commands.handelInputValues('SOCKET_IO_PORT', watchConfig.socket, 3030))
+      .option('-o,--output-type [n]', 'Output type', Commands.handelInputValues('OUTPUT_TYPE', watchConfig.outputType, 'terminal'))
```
Comment (MostafaBalata) [inline src/command.js]: '3030' shouldn't be hard coded here, it should be config variable.

Comment (MostafaBalata) [reply src/command.js]: Do we use `SOCKET_IO`?

File: src/command.js [hunk 7]
```diff
@@ -2,74 +2,115 @@ import program from 'commander';
...
+      .option('-e,--access-token [n]', 'etherscan accssess token', Commands.handelInputValues('ACCESS_TOKEN', watchConfig.accessToken, ''))
+      .parse(process.argv);
 
-  program.parse(process.argv);
-  if (typeof program === 'undefined') { throw new Error('No args are specifed in the command or in the .watch.yml file'); }
+    if (typeof program === 'undefined') { throw new Error('No args are specifed in the command or in the .watch.yml file'); }
+    // TODO: add validation to all parameters
```
Comment (MostafaBalata) [inline src/command.js]: add validation to all parameters

File: src/command.js (hunk 2 above)
Comment (MostafaBalata) [reply src/command.js]: are you sure `handle` is good name? I would use `get` instead.

File: src/etherscan.js [hunk 8]
```diff
@@ -32,7 +32,7 @@ const validateResponse = (response) => {
  *   @return object
  */
 
-const scrapeABI = async (address) => {
+const scrapeABI = async (address, accessToken) => {
```
Comment (MostafaBalata) [inline src/etherscan.js]: Why did you change the `accessToken`  to be function parameter?

File: src/index.js [hunk 9]
```diff
@@ -35,53 +34,58 @@ const transactionHandler = async (transaction) => {
...
-  const lastBlockNumber = await bluebird.promisify(web3.eth.getBlockNumber)();
-
-  const { from, to, addresses, quickMode,
-    lastBlockNumberFilePath } = command(watchingConfigPath, lastBlockNumber);
-  logger.debug('Start process');
+  try {
+    // @TODO: find a solution for variable handeling
+    variableClass = await initCommands();
+    logger = loggerInit(variableClass.logLevel);
+    console.log(logger);
```
Comment (MostafaBalata) [inline src/index.js]: You shouldn't use console

File: src/index.js [hunk 10]
```diff
@@ -35,53 +34,58 @@ const transactionHandler = async (transaction) => {
...
+    const jsonRpc = new JsonRpc(variableClass.addresses,
+      variableClass.from, variableClass.to,
+      variableClass.lastBlockNumberFilePath, logger,
+      transactionHandler);
 
-    await jsonRpc.scanBlocks(quickMode);
+    await jsonRpc.scanBlocks(variableClass.quickMode);
     logger.info('Finish scanning all the blocks');
   } catch (e) {
+    console.log(e);
```
Comment (MostafaBalata) [inline src/index.js]: console?

File: src/index.js [hunk 11]
```diff
@@ -35,53 +34,58 @@ const transactionHandler = async (transaction) => {
...
     logError(e);
   }
 };
 
 main().catch((e) => {
+  console.log(e);
```
Comment (MostafaBalata) [inline src/index.js]: console?

File: src/jsonrpc.js [hunk 12]
```diff
@@ -15,7 +15,8 @@ export default class JsonRpc {
    * @param int toBlock
    * @param function callback
    */
-  constructor(addresses, fromBlock, toBlock, lastBlockNumberFilePath = null, callback = null) {
+  constructor(addresses, fromBlock, toBlock,
+    lastBlockNumberFilePath = null, logger, callback = null) {
```
Comment (MostafaBalata) [inline src/jsonrpc.js]: why logger should be parameter?

Comment (MostafaBalata) [reply src/jsonrpc.js]: the order of the variables is not correct

File: src/jsonrpc.js [hunk 13]
```diff
@@ -25,8 +26,9 @@ export default class JsonRpc {
     this.toBlock = toBlock !== defaultBlockNumber ? toBlock : null;
     this.web3Instance = web3.eth;
     this.callback = callback;
+    this.logger = logger;
```
Comment (MostafaBalata) [inline src/jsonrpc.js]: WHY?

File: src/logger.js [hunk 14]
```diff
@@ -1,25 +1,24 @@
 import winston from 'winston';
-import { getEnv } from './config';
 
-const loggerConsoleOptions = {
-  timestamp: false,
-  level: getEnv('LOG_LEVEL'),
-  colorize: false,
-  formatter: options => `${options.message}`,
+const logger = (logLevel) => {
```
Comment (MostafaBalata) [inline src/logger.js]: Why do you think importing config variables is bad, and make every thing parameter is better?

File: src/logger.js [hunk 15]
```diff
@@ -1,25 +1,24 @@
...
 /**
  * This will print out the error as json formatted
- * @param {*} error 
- * @param {*} customMessage 
+ * @param {*} error
+ * @param {*} customMessage
  */
-export const logError = (error, customMessage = null, isStack = true) => {
-  switch (getEnv('OUTPUT_TYPE')) {
+export const logError = (error, customMessage = null, isStack = true, outputType) => {
```
Comment (MostafaBalata) [inline src/logger.js]: bad order for variables.

File: src/web3/web3Provider.js [hunk 16]
```diff
@@ -6,7 +6,7 @@ const getWeb3 = () => {
   if (typeof web3 !== 'undefined') {
     web3 = new Web3(web3.currentProvider);
   } else {
-    // set the provider you want from Web3.providers
+    // TODO: remove this dependancy on ENV
```
Comment (MostafaBalata) [inline src/web3/web3Provider.js]: what does that mean?

File: src/command.js (hunk 2 above)
Comment (Moejoe90) [reply src/command.js]: By using commander the priority will be for 
1- CLI
2- ENV
3- .watch
4- Default value 

the reason for this is to take off the reliance on node as a server. We did a mistake assuming this application will be run from a server point of view. If you want a CLI tool + server tool then there is no running away from giving the user the ability to utilize anything be that a configuration file, env variable or CLI directly. 
In order to use the tool as fully CLI then we need to get rid of relying on Env variables. 

Regarding `handle` to `get` i will update now

File: .env.example (hunk 3 above)
Comment (Moejoe90) [reply .env.example]: Now with this change everything can be an env variable, this is better for system admins to choose what settings are bet for them

File: src/command.js (hunk 4 above)
Comment (Moejoe90) [reply src/command.js]: Fair enough

File: src/command.js (hunk 5 above)
Comment (Moejoe90) [reply src/command.js]: Also fair enough

File: src/command.js (hunk 6 above)
Comment (Moejoe90) [reply src/command.js]: Do we use socket io when dealing with graylog? i just used what i found in the env file if that's not the case then it should have been removed sometime ago

File: src/command.js (hunk 1 above)
Comment (Moejoe90) [reply src/command.js]: It was a solution to the problem, with a class all variables declared as this have `getters` automatically. do you have a better solution?

File: src/command.js [hunk 17]
```diff
@@ -1,75 +1,135 @@
...
-  program.parse(process.argv);
-  if (typeof program === 'undefined') { throw new Error('No args are specifed in the command or in the .watch.yml file'); }
+/**
+ * Validate address string based on web3.js function isAddress
+ * then default value. should throw incase of error
+ * @param {array} addresses
+ * @return addresses
+ */
+const validateParamter = (parameter, errMsg) => {
+  if (!parameter || !parameter.length > 0) {
```
Comment (MostafaBalata) [inline src/command.js]: Is `!parameter.length > 0` the same as `parameter.length <= 0`?

File: src/command.js [hunk 18]
```diff
@@ -1,75 +1,135 @@
...
+    if (!isPathExist(saveStatePath)) { fs.mkdirSync(saveStatePath); }
     if (!isPathExist(lastBlockNumberFilePath)) {
       fs.writeFileSync(lastBlockNumberFilePath, JSON.stringify({ blockNumber: from }));
-    } else {
-      const lastBlockNumberJson = JSON.parse(fs.readFileSync(lastBlockNumberFilePath,
-        { encoding: 'utf8' }));
-      from = lastBlockNumberJson.blockNumber;
+      return from;
     }
+    return (JSON.parse(fs.readFileSync(lastBlockNumberFilePath, { encoding: 'utf8' }))).blockNumber;
```
Comment (MostafaBalata) [inline src/command.js]: `return JSON.parse(fs.readFileSync(lastBlockNumberFilePath, { encoding: 'utf8' })).blockNumber` isn't same?

File: src/command.js [hunk 19]
```diff
@@ -1,75 +1,135 @@
...
-import { defaultBlockNumber } from './config';
-import { isAddress, validateBlockNumber } from './web3/utils';
+import { defaultBlockNumber, watchingConfigPath, getEnv, saveStateFileName } from './config';
 import { isPathExist } from './utils';
 
+const rpcErrorMsg = '-n or --nodeAddress is required';
+const addressErrorMsg = '-a or --address is required';
+const quickModeErrorMsg = '-q or --quick-mode shouldn\'t have any value';
+const loggerErrorMsg = 'Logger is not correct type';
+const noArgsErrorMsg = 'No args are specifed in the command or in the .watch.yml file';
```
Comment (MostafaBalata) [inline src/command.js]: specified

File: src/command.js [hunk 20]
```diff
@@ -1,75 +1,135 @@
...
+ * @param {integer} toBlock
+ *
+ */
+const validateBlockbySize = (fromBlock, toBlock) => {
+  if (toBlock !== defaultBlockNumber && fromBlock > toBlock) { throw new Error(`From "${fromBlock}" shouldn't be larger than "${toBlock}"`); }
+  return fromBlock;
+};
 
+/**
+ * Create saveState file if it doesn't exist, read last Blockwritten in file if it exists
```
Comment (MostafaBalata) [inline src/command.js]: Block written

File: src/command.js [hunk 21]
```diff
@@ -1,75 +1,135 @@
...
-  const quickMode = program.quick;
-  const saveStatePath = program.saveState;
-  let lastBlockNumberFilePath = null;
+/**
+ * Should throw if starting block is bigger than endblock while end block is not default
+ * @param {integer} fromBlock
+ * @param {integer} toBlock
+ *
+ */
+const validateBlockbySize = (fromBlock, toBlock) => {
```
Comment (MostafaBalata) [inline src/command.js]: validateBlockBySize

File: src/index.js [hunk 22]
```diff
@@ -50,29 +47,28 @@ const transactionHandler = async (transaction) => {
       logger.error(`txHash: ${transaction.hash} ${error.message}`);
     }
   }
-  output({ transaction, decodedInputDataResult, decodedLogs }, getEnv('OUTPUT_TYPE'));
+  output({ transaction, decodedInputDataResult, decodedLogs }, command().outputType);
```
Comment (MostafaBalata) [inline src/index.js]: why you are not using `getCommandVars`

File: src/web3/utils.js [hunk 23]
```diff
@@ -1,28 +1,36 @@
...
 /**
  * Check if block number is correct by passing the last block number
  * in blockchain and the targeted block number
  * @param lastBlockNumber
  * @param blockNumber
  */
-export const validateBlockNumber = (lastBlockNumber, blockNumber) => {
-  if (isNaN(blockNumber) || lastBlockNumber < blockNumber) {
-    throw new Error(`${blockNumber} ${lastBlockNumber} is not valid block number`);
+export const validateBlockNumber = async (blockNumber) => {
```
Comment (MostafaBalata) [inline src/web3/utils.js]: you converted it into async but you didn't await it everywhere.

File: src/web3/web3Provider.js [hunk 24]
```diff
@@ -1,15 +1,15 @@
...
 
 let web3;
-const getWeb3 = () => {
+export const getWeb3 = () => {
   if (typeof web3 !== 'undefined') {
     web3 = new Web3(web3.currentProvider);
   } else {
-    // set the provider you want from Web3.providers
-    web3 = new Web3(new Web3.providers.HttpProvider(getEnv('RPC_URL')));
+    // TODO: remove this dependancy on ENV
```
Comment (MostafaBalata) [inline src/web3/web3Provider.js]: what is this TODO about?

File: .env.example [hunk 25]
```diff
@@ -1,4 +1,9 @@
-LOG_LEVEL= "info"
-RPC_URL = http://localhost:8545
-OUTPUT_TYPE = "graylog"
-ACCESS_TOKEN = ""
\ No newline at end of file
+LOG_LEVEL= 'debug'
+SOCKET_IO_PORT = 3030
```
Comment (MostafaBalata) [inline .env.example]: we are not using sockets

File: .env.example [hunk 26]
```diff
@@ -1,4 +1,9 @@
...
+RPC_URL = "http://localhost:8545"
+OUTPUT_TYPE = "terminal"
+ACCESS_TOKEN = ""
+QUICK_MODE = true
+TO_BLOCK
+FROM_BLOCK
+ADDRESS
```
Comment (MostafaBalata) [inline .env.example]: Give them any values
`TO_BLOCK`
`FROM_BLOCK`
`ADDRESS`

File: null/last-block-number.json [hunk 27]
```diff
@@ -0,0 +1 @@
+{"blockNumber":4240720}
```
Comment (MostafaBalata) [inline null/last-block-number.json]: What is this?

File: src/web3/utils.js [hunk 28]
```diff
@@ -1,28 +1,36 @@
...
+  if (blockNumber === defaultBlockNumber) return defaultBlockNumber;
+  const lastBlockNumber = await getLastBlock();
+  if (isNaN(blockNumber) || await lastBlockNumber < blockNumber) {
```
Comment (MostafaBalata) [inline src/web3/utils.js]: why are you awaiting awaited value?

**Comment by Moejoe90** [comment]
I completely agree with removing `copyfile.sh`. I assumed it must have been used with docker compose since that is not the case it is removed.
It was this copyfile.sh was already taken of tests and now taken of the start
